Mainnet/HashBattle/HashBattle/Arena.cs

```diff
@@ -0,0 +1,194 @@
+    /// </summary>
+    public bool EnterBattle(ulong battleId, uint userindex)
+    {
+        var battle = GetBattle(battleId);
+
+        Assert(battle.Winner == Address.Zero, "Battle ended.");
+
+        Assert(battle.Fee == Message.Value, "Battle amount is not matching.");
+
+        var user = GetUser(battleId, Message.Sender);
+
+        Assert(!user.ScoreSubmitted, "The user already submitted score.");
+
+        user.Address = Message.Sender;
+
+        SetUser(battleId, Message.Sender, user);
+
+        battle.Users.SetValue(user.Address, userindex);
+        SetBattle(battleId, battle);
+
+        Log(battle);
+        return true;
+    }
+
+    /// <summary>
+    /// 4 different user will end the battle and submit the score
+    /// </summary>
+    public bool EndBattle(Address userAddress, ulong battleId, uint score, bool IsBattleOver)
+    {
+        Assert(Message.Sender == BattleOwner, "Only battle owner can end game.");
+
+        var battle = GetBattle(battleId);
+
+        Assert(battle.Winner == Address.Zero, "Battle ended.");
+
+        var user = GetUser(battleId, userAddress);
+
+        Assert(!user.ScoreSubmitted, "The user already submitted score.");
+
+        user.Score = score;
+        user.ScoreSubmitted = true;
+
+        SetUser(battleId, userAddress, user);
+
+        if (IsBattleOver)
+            ProcessWinner(battle);
+
+        Log(user);
+        return true;
+    }
+
+    /// <summary>
+    /// Get winner address
+    /// </summary>
+    public Address GetWinner(ulong battleId)
+    {
+        var battle = GetBattle(battleId);
+        Log(battle);
+        return battle.Winner;
+    }
+
+    /// <summary>
+    /// Process winner when all user scores are submitted
+    /// </summary>
+    private void ProcessWinner(BattleMain battle)
+    {
+        if (battle.Users.Length <= 4)
+        {
+            foreach (Address userAddress in battle.Users)
+            {
+                var user = GetUser(battle.BattleId, userAddress);
+                if (!user.ScoreSubmitted)
+                    return;
+            }
+        }
+        uint winnerIndex = GetWinnerIndex(battle.BattleId, battle.Users);
+        if (battle.Winner == Address.Zero)
+        {
+            battle.Winner = battle.Users[winnerIndex];
+            SetBattle(battle.BattleId, battle);
+            ProcessPrize(battle.BattleId);
+        }
+    }
+
+    /// <summary>
+    /// Get winner user index from battle users
+    /// </summary>
+    private uint GetWinnerIndex(ulong battleid, Address[] users)
+    {
+        uint winningScore = 0;
+        uint winningScoreIndex = 0;
+        for (uint i = 0; i < users.Length; i++)
+        {
+            var user = GetUser(battleid, users[i]);
+            if (user.Score > winningScore)
+            {
+                winningScore = user.Score;
+                winningScoreIndex = i;
+            }
+        }
+        return winningScoreIndex;
+    }
```

Comment:
Do you need to consider what happens in the case of a tie? As it stands, the user with the lowest index will be determined as the winner.

Reply:
Last indexed user will be winner in this case if users have same score. Just fyi